src/test/shell/bazel/target_compatible_with_test.sh

```diff
@@ -0,0 +1,690 @@
+    target_compatible_with = [
+        ":foo1",
+    ],
+)
+
+genrule(
+    name = "use_host_tool",
+    outs = ["host_tool_message.txt"],
+    cmd = "\$(location :host_tool) >> \$(OUTS)",
+    tools = [":host_tool"],
+    target_compatible_with = [
+        ":foo2",
+    ],
+)
+EOT
+
+  cd target_skipping || fail "couldn't cd into workspace"
+
+  # Run with :foo1 in the host platform, but with :foo2 in the target platform.
+  # Building the host tool should fail because it's incompatible with the
+  # target platform.
+  if bazel build --show_result=10 --host_platform=@//:foo1_bar1_platform \
+    --platforms=@//:foo2_bar1_platform //:host_tool &> "${TEST_log}"; then
+    fail "Bazel passed unexpectedly."
+  fi
+  expect_log 'ERROR: Target //:host_tool is incompatible and cannot be built, but was explicitly requested'
+  expect_log 'FAILED: Build did NOT complete successfully'
+
+  # Run with :foo1 in the host platform, but with :foo2 in the target platform.
+  # This should work fine because we're not asking for any constraints to be
+  # violated.
+  if ! bazel build --host_platform=@//:foo1_bar2_platform \
+    --platforms=@//:foo2_bar1_platform //:host_tool_message.txt \
+    &> "${TEST_log}"; then
+    fail "Bazel failed unexpectedly."
+  fi
+  expect_log ' bazel-bin/host_tool_message.txt$'
+  expect_log ' Build completed successfully, '
+
+  # Make sure that the contents of the file are what we expect.
+  cp bazel-bin/host_tool_message.txt "${TEST_log}"
+  expect_log '^Hello World$'
+}
+
+# Validates that incompatible targets provide appropriate errors when queried
+# explicitly or skipped when queried via a glob.
+function test_queries() {
+  cat >> target_skipping/BUILD <<EOT
+genrule(
+    name = "genrule_foo1",
+    target_compatible_with = [":foo1"],
+    outs = ["foo1.sh"],
+    cmd = "echo 'exit 0' > \$(OUTS)",
+)
+
+sh_binary(
+    name = "sh_foo1",
+    srcs = ["foo1.sh"],
+)
+EOT
+
+  cd target_skipping || fail "couldn't cd into workspace"
+
+  if ! bazel query 'deps(//:sh_foo1)' &> "${TEST_log}"; then
+    fail "Bazel query failed unexpectedly."
+  fi
+  expect_log '^//:sh_foo1$'
+  expect_log '^//:genrule_foo1$'
+
+  # Run a cquery on a target that is compatible. This should pass.
+  if ! bazel cquery --host_platform=@//:foo3_platform \
+    --platforms=@//:foo3_platform \
+    'deps(//:some_foo3_target)' &> "${TEST_log}"; then
+    fail "Bazel cquery failed unexpectedly."
+  fi
+
+  # Run a cquery with a glob. This should only pick up compatible targets and
+  # should therefore pass.
+  if ! bazel cquery --host_platform=@//:foo3_platform \
+    --platforms=@//:foo3_platform 'deps(//...)' &> "${TEST_log}"; then
+    fail "Bazel cquery failed unexpectedly."
+  fi
+  expect_log '^//:some_foo3_target '
+
+  # TODO(philsc): Should the targets show up here? I would have expected them
+  # to be omitted. In other words, I expected to be able to use expect_not_log.
+  expect_log '^//:sh_foo1 '
+  expect_log '^//:genrule_foo1 '
```

Comment:
@gregestren, what do you expect the behaviour of `bazel cquery` to be? I assumed that querying `//...` should skip incompatible targets and not list them, but I really haven't looked into the implementation at all. Right now incompatible targets do show up in `deps(//...)`.

Explicitly querying on an incompatible target generates the same error as explicitly building an incompatible target.

Reply:
This behavior makes sense. 

For explicit querying, I'd expect failure because cquery works by first running the analysis phase, then looking at the graph that got produced from that. So if the analysis phase fails with an error that automatically means cquery fails.

For implicit queries, as long as the analysis phase doesn't fail, then it just passes the dependency graph to cquery, which does its own traversal. cquery doesn't particularly look for incompatibility (it's following a completely separate process from the BuildTool logic that you're affecting here). If a dep is part of of the graph it reports it.

I'd probably lean toward this behavior for now on the account of it being more inclusive. cquery is often used to figure out things like, "what files do I need in order to support `//...`"?). So overapproximation is a good thing. I'd support either a followup change or a flag to toggle this only after we get some hands-on experience showing the need.

Reply:
Okay. That makes sense. Thank you!